### PR TITLE
JavaScript Capstone Project

### PR DESCRIPTION
**Pull request by Rhaegar121**
In this PR, we have completed the following tasks...

1. Used the TVmaze API
2. Displayed the list of movies on the home page
3. Displayed comment popup with selected movie details
4. Displayed the number of likes for each item on the home page using the involvement API
5. Displayed comments for a given movie on the comments pop-up
6. Created two features for adding new likes and comment
7. Added counters for the number of movies and comments
8. Added tests for the counters
9. Recorded the 5 min video for the project description

### CODE REVIEW

File: src/modules/movieCounter.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+const movieCount = (movie) => movie.length;
```
Comment (huseyinbiyik) [inline src/modules/movieCounter.js]: Good job with comment counter. 👍  But let's follow the requirements and use same counting method for movies also.
>Even if the API gives you these numbers, you will create a specific function to calculate these numbers in each page. 
Screenshot 2023-02-08 at 9 11 36 PM

File: src/_test_/commentCounter.test.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/** @jest-environment jsdom */
```
Comment (huseyinbiyik) [inline src/_test_/commentCounter.test.js:1]: Good job with adding tests but you need to mock the DOM environment instead of a giving an array. Kindly check the code example below and fix your tests.
```
const { JSDOM } = require('jsdom');

describe('CommentCounter', () => {
  let CommentsCounter;
  let dom;
  
  beforeEach(() => {
    dom = new JSDOM(`
      <!DOCTYPE html>
      <html>
        <body>
          <div id="comments">
            <p class="comment">Comment 1</p>
            <p class="comment">Comment 2</p>
            <p class="comment">Comment 3</p>
          </div>
        </body>
      </html>
    `);
    global.document = dom.window.document;
    global.window = dom.window;
    CommentsCounter = require('../CommentsCounter');
  });
  
  it('counts the number of elements with class "comment"', () => {
    const count = CommentsCounter();
    expect(count).toEqual(3);
  });
});
```

File: src/_test_/movieCounter.test.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/** @jest-environment jsdom */
```
Comment (huseyinbiyik) [inline src/_test_/movieCounter.test.js:1]: Good job adding test for the movieCounter. After fixing your movieCounter function, you need to fix your tests here and add a mock environment. You can use follow the example I shared for the commentsCounter. 👍

File: dist/index.html [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+            <span>Movies</span>
+            <span>&</span>
+            <span>TV Series</span>
+        </div>
+        <div id="link">
+            <a href="#">Home</a>
+            <a href="#">About</a>
+            <a href="#">Contact</a>
+        </div>
+    </header>
```
Comment (huseyinbiyik) [inline dist/index.html:23]: Good job with customizing the concept 💯  But let's follow the layout provided. Please put the elements exactly how it looks on the design. 👍 
Screenshot 2023-02-08 at 9 34 25 PM
Screenshot 2023-02-08 at 9 34 55 PM

**Review by huseyinbiyik — CHANGES_REQUESTED** [review]
Hi @Victor-chiemerie and @Rhaegar121 🐉 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
Almost there

## Highlights
- Functions were separated into modules 👌 
- Async functions were used 👍 
- API transactions were implemented properly 👌 
- Linter set up well and checks pass 🟢
- Git flow followed 👌
- PR description is descriptive 📝

## Required Changes ♻️
_Check the comments under the review._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/modules/movieCounter.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+const movieCount = (movie) => {
+  let count = 0;
+  for (let i = 0; i < movie.length; i += 1) {
+    count += 1;
+  } return count;
+};
```
Comment (amejid) [inline src/modules/movieCounter.js]: - [ ] Amazing job with all the functionalities and design. However, the counter is supposed to count HTML elements on the page. You are still rendering the length of data retrieved from the API. I think you are using a div with a class name `container` for your movies, so you can use `document.querySelectorAll('.container')` to get all the HTML nodes with that class name. Kindly perform the counting on the HTML nodes. ✍️
- [ ] [OPTIONAL] I highly suggest you do the same thing for the comment counter. 😎️

File: src/_test_/movieCounter.test.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
```
Comment (amejid) [inline src/_test_/movieCounter.test.js:1]: - [ ] Since the counters are supposed to count HTML elements in the DOM, the tests are also supposed to check if the counters are correctly working. Kindly use the example test suggested by the previous reviewer to achieve adequate results. 🧐️

File: src/_test_/commentCounter.test.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
```
Comment (amejid) [inline src/_test_/commentCounter.test.js:1]: - [ ] Please do the same type of tests suggested for the movie counter here as mentioned by the previous reviewer. 🥸️

File: src/modules/displayData.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    <button>Reservations</button>`;
+    home.appendChild(dataCard);
+
+    // updating likes on the home page
+    const likeBtn = dataCard.querySelector('.fa-heart');
+    likeBtn.onclick = () => {
+      const like = dataCard.querySelector('#like');
+      pushLike(data.id);
+      like.innerHTML = `${likes[0].likes + 1} Likes`;
+    };
```
Comment (amejid) [inline src/modules/displayData.js]: - [x] [OPTIONAL] I noticed that when clicking the like button on a movie with `0` likes it throws an error and does not update the interface. I highly suggest you take a look into it.

**Review by amejid — CHANGES_REQUESTED** [review]
Hi team, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



 ## Highlights
- Linter checks passing :heavy_check_mark: 
- Good UI design :heavy_check_mark: 
- Descriptive commits :heavy_check_mark: 
- Good collaboration history :heavy_check_mark: 
 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/modules/movieCounter.js (hunk 5 above)
Comment (Rhaegar121) [reply src/modules/movieCounter.js]: Hi, I changed the movie counter to count HTML elements on the page but for the comment counter, how can I select the dom element for a specific comment container because its id is dynamically created? Can you explain this to me?

Comment (amejid) [reply src/modules/movieCounter.js]: Hi @Rhaegar121, you can call the comments counter function after the comments are rendered on the screen. You can give the specific comment a class name to assist you in the counting process.

Screenshot from 2023-02-09 12-06-58

Comment (amejid) [reply src/modules/movieCounter.js]: The counter could look like this.

image

Comment (Rhaegar121) [reply src/modules/movieCounter.js]: Hi @amejid ,

Yeah I have tried the querySelectorAll method but it didn't work. The output show only 1 for all the movies' comment. I think I have to select the particular comment section but I don't know how to do or it is possible to do.

Screenshot 2023-02-09 155203
Screenshot (55)
Screenshot (56)

Comment (Rhaegar121) [reply src/modules/movieCounter.js]: If you have any references or examples, I'll be happy to learn from them. 😊

Comment (amejid) [reply src/modules/movieCounter.js]: I hope the issue cleared up after the call.

Comment (Rhaegar121) [reply src/modules/movieCounter.js]: Yes, Thanks for the help bro. You save me a lot of time.

File: README.md [hunk 9]
```diff
@@ -51,184 +17,129 @@ After you're finished please remove all the comments and instructions!
...
 
 <p align="right">(<a href="#readme-top">back to top</a>)</p>
 
 <!-- LIVE DEMO -->
 
 ## 🚀 Live Demo <a name="live-demo"></a>
 
-> Add a link to your deployed project.
+- live demo link will appear when ready
+
```
Comment (OyePriscilla) [inline README.md:73]: - [optional] I think it will be nice to deploy your app online, this will ensure that your README file is professional to people visiting your repo. 👍

File: README.md [hunk 10]
```diff
@@ -51,184 +17,129 @@ After you're finished please remove all the comments and instructions!
...
-- GitHub: [@githubhandle](https://github.com/githubhandle)
-- Twitter: [@twitterhandle](https://twitter.com/twitterhandle)
-- LinkedIn: [LinkedIn](https://linkedin.com/in/linkedinhandle)
+- GitHub: [@githubhandle](https://github.com/Victor-chiemerie)
 
 <p align="right">(<a href="#readme-top">back to top</a>)</p>
 
 <!-- FUTURE FEATURES -->
 
 ## 🔭 Future Features <a name="future-features"></a>
```
Comment (OyePriscilla) [inline README.md:142]: - [optional] Please kindly add one or more future feature(s) to your README file according to the project requirement. ℹ️

File: src/_test_/commentCounter.test.js [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+    <div class="comment-text"></div>
+    <div class="comment-text"></div>
+    <div class="comment-text"></div>
+    <div class="comment-text"></div>
+  </div>`;
+
+  test('number of comments should be 3', async () => {
+    expect(countComments()).toEqual(4);
+  });
+});
```
Comment (OyePriscilla) [inline src/_test_/commentCounter.test.js:16]: - [optional] It's awesome implementing the changes required by previous reviewers, but I think you would have test for all edge cases as well to make your test functions robust (perhaps a scenario in which there are 0 items on the page, counter function should return 0). 👍

File: src/style.css [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+  padding: 30px 50px;
+  gap: 15px;
+  align-items: center;
+  flex-direction: column;
+  display: flex;
+  position: relative;
+  overflow: hidden;
+  overflow-y: auto;
+  height: 100vh;
+}
```
Comment (OyePriscilla) [inline src/style.css:181]: - [optional] The contrast between the **background** and the **text** of this container is too close, kindly adjust this for great User Experience. 👍  

## Your Design
DSC_7491

**Review by OyePriscilla — APPROVED** [review]
Hi @Rhaegar121 and @Victor-chiemerie 😄, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
✔️Git flow observed 👍
✔️App worked properly 👍
✔️Made required changes 👍

## Optional suggestions

- [optional] The UI/UX design of your app is not responsive, It would be great if you can work on the mobile version as well for a great UI/UX. 👍 

> **Note: All these are optional because they were not highlighted by the previous reviewer(s). I would strongly advice you to implement them to have a great web app** 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
